File: javascript/json_batch_processor.js
#!/usr/bin/env node
// json_batch_processor.py — portiert nach javascript
// Quelle: python, OpenClaw@gateway1:skills/json-utils/scripts/json_batch_processor.py
// auch in: OpenClaw@gateway2:skills/json-utils/scripts/json_batch_processor.py
// Erzeugt: 2026-08-19 durch ABSTRACTIONS_MANAGER.py

/**
 * Batch JSON Processor - Verarbeitet mehrere JSON-Dateien oder JSON-Lines (NDJSON).
 */

const fs = require('fs');
const path = require('path');
const { program } = require('commander');

// Simuliere einige Python-ähnliche Funktionen
function* readJsonl(filePath) {
    /**
     * Liest JSON-Lines (NDJSON) Datei Zeile für Zeile.
     *
     * @param {string} filePath - Pfad zur .jsonl oder .ndjson Datei
     * @yields {Object|BatchResult} - Geparste JSON-Objekte oder Fehler
     */
    const content = fs.readFileSync(filePath, 'utf-8');
    const lines = content.split('\n');
    
    for (let lineNum = 0; lineNum < lines.length; lineNum++) {
        const line = lines[lineNum].trim();
        if (!line) continue;
        
        try {
            yield JSON.parse(line);
        } catch (e) {
            yield new BatchResult(
                lineNum + 1,
                `${filePath}:${lineNum + 1}`,
                false,
                null,
                `JSON decode error: ${e.message}`
            );
        }
    }
}

class BatchResult {
    /** Ergebnis einer Batch-Verarbeitung. */
    
    constructor(index, source, success, data = null, error = null) {
        this.index = index;
        this.source = source;
        this.success = success;
        this.data = data;
        this.error = error;
    }
    
    toDict() {
        return {
            index: this.index,
            source: this.source,
            success: this.success,
            data: this.data,
            error: this.error
        };
    }
}

class JSONProcessingError extends Error {
    constructor(message) {
        super(message);
        this.name = 'JSONProcessingError';
    }
}

function parseJson(content, repair = true) {
    /**
     * Parst JSON-Inhalt mit optionaler Reparatur.
     * 
     * @param {string} content - JSON-String
     * @param {boolean} repair - Ob Reparatur versucht werden soll
     * @returns {Object} - Geparstes JSON
     */
    try {
        return JSON.parse(content);
    } catch (e) {
        if (repair) {
            // Versuche einfache Reparaturen
            let repaired = content.replace(/,\s*}/g, '}').replace(/,\s*\]/g, ']');
            try {
                return JSON.parse(repaired);
            } catch (repairError) {
                throw new JSONProcessingError(`JSON parsing failed: ${e.message}`);
            }
        } else {
            throw new JSONProcessingError(`JSON parsing failed: ${e.message}`);
        }
    }
}

function parseAndValidate(content, validateModel, repair = true) {
    /**
     * Parst und validiert JSON gegen ein Model.
     * 
     * @param {string} content - JSON-String
     * @param {Function} validateModel - Validierungsmodel (Konstruktor)
     * @param {boolean} repair - Ob Reparatur versucht werden soll
     * @returns {Object} - Validiertes JSON
     */
    const data = parseJson(content, repair);
    // In JS können wir keine echte Pydantic-Validierung machen,
    // aber wir simulieren es durch einen Konstruktoraufruf
    if (validateModel && typeof validateModel === 'function') {
        try {
            return new validateModel(data);
        } catch (validationError) {
            throw new JSONProcessingError(`Validation failed: ${validationError.message}`);
        }
    }
    return data;
}

async function processBatch(inputs, processor, maxWorkers = 4) {
    /**
     * Verarbeitet eine Liste von Inputs parallel.
     *
     * @param {Array} inputs - Liste der zu verarbeitenden Eingaben
     * @param {Function} processor - Funktion, die (input, index) -> BatchResult zurückgibt
     * @param {number} maxWorkers - Anzahl paralleler Worker
     * @returns {Promise<Array>} - Liste der BatchResult-Objekte
     */
    const results = [];
    const workerPromises = [];
    const workers = Math.min(maxWorkers, inputs.length);
    
    // Chunk inputs for workers
    const chunks = Array.from({ length: workers }, (_, i) => 
        inputs.filter((_, idx) => idx % workers === i)
    );
    
    for (let workerIdx = 0; workerIdx < workers; workerIdx++) {
        const chunk = chunks[workerIdx];
        workerPromises.push(
            Promise.all(
                chunk.map(async (inp, localIdx) => {
                    const globalIdx = inputs.indexOf(inp);
                    try {
                        return await processor(inp, globalIdx);
                    } catch (e) {
                        return new BatchResult(
                            globalIdx,
                            String(inp),
                            false,
                            null,
                            `Unexpected error: ${e.message}`
                        );
                    }
                })
            )
        );
    }
    
    const workerResults = await Promise.all(workerPromises);
    for (const workerResult of workerResults) {
        results.push(...workerResult);
    }
    
    // Sortiere nach Index
    results.sort((a, b) => a.index - b.index);
    return results;
}

async function processFileBatch(filePaths, repair = true, validateModel = null, maxWorkers = 4) {
    /**
     * Verarbeitet mehrere JSON-Dateien im Batch.
     *
     * @param {Array<string>} filePaths - Liste der Datei-Pfade
     * @param {boolean} repair - Ob JSON-Reparatur angewendet werden soll
     * @param {Function|null} validateModel - Optionales Validierungsmodell
     * @param {number} maxWorkers - Parallele Worker
     * @returns {Promise<Array>} - Liste der BatchResult-Objekte
     */
    async function processor(filePath, idx) {
        try {
            const content = fs.readFileSync(filePath, 'utf-8');
            
            let data;
            if (validateModel) {
                data = parseAndValidate(content, validateModel, repair);
            } else {
                data = parseJson(content, repair);
            }
            
            return new BatchResult(
                idx,
                filePath,
                true,
                data
            );
        } catch (e) {
            if (e instanceof JSONProcessingError) {
                return new BatchResult(
                    idx,
                    filePath,
                    false,
                    null,
                    e.message
                );
            } else {
                return new BatchResult(
                    idx,
                    filePath,
                    false,
                    null,
                    `${e.constructor.name}: ${e.message}`
                );
            }
        }
    }
    
    return await processBatch(filePaths, processor, maxWorkers);
}

function processJsonlFile(filePath, repair = true, validateModel = null) {
    /**
     * Verarbeitet eine JSON-Lines Datei.
     *
     * @param {string} filePath - Pfad zur .jsonl Datei
     * @param {boolean} repair - Ob Reparatur angewendet werden soll
     * @param {Function|null} validateModel - Optionales Validierungsmodell
     * @returns {Array} - Liste der BatchResult-Objekte
     */
    const results = [];
    const lines = fs.readFileSync(filePath, 'utf-8').split('\n');
    
    for (let lineNum = 0; lineNum < lines.length; lineNum++) {
        const line = lines[lineNum].trim();
        if (!line) continue;
        
        try {
            let data;
            if (validateModel) {
                data = parseAndValidate(line, validateModel, repair);
            } else {
                data = parseJson(line, repair);
            }
            
            results.push(new BatchResult(
                lineNum + 1,
                `${filePath}:${lineNum + 1}`,
                true,
                data
            ));
        } catch (e) {
            if (e instanceof JSONProcessingError) {
                results.push(new BatchResult(
                    lineNum + 1,
                    `${filePath}:${lineNum + 1}`,
                    false,
                    null,
                    e.message
                ));
            }
        }
    }
    
    return results;
}

function writeJsonl(results, outputPath, onlySuccessful = true) {
    /**
     * Schreibt BatchResult-Liste als JSON-Lines.
     *
     * @param {Array} results - Liste der Ergebnisse
     * @param {string} outputPath - Ausgabedatei
     * @param {boolean} onlySuccessful - Nur erfolgreiche Ergebnisse schreiben
     */
    const fd = fs.openSync(outputPath, 'w');
    
    for (const result of results) {
        if (onlySuccessful && !result.success) {
            continue;
        }
        fs.writeSync(fd, JSON.stringify(result.toDict()) + '\n');
    }
    
    fs.closeSync(fd);
}

async function main() {
    program
        .description('Batch JSON Processor')
        .argument('<inputs...>', 'JSON files to process')
        .option('--jsonl, -l', 'Treat inputs as JSON-Lines files', false)
        .option('--repair, -r', 'Enable JSON repair', true)
        .option('--workers, -w <number>', 'Parallel workers', parseInt, 4)
        .option('--output, -o <file>', 'Output JSON-Lines file')
        .option('--summary, -s', 'Show summary only', false)
        .action(async (inputs, options) => {
            let allResults = [];
            
            if (options.jsonl) {
                // JSON-Lines Modus
                for (const inputPath of inputs) {
                    const results = processJsonlFile(inputPath, options.repair);
                    allResults = allResults.concat(results);
                }
            } else {
                // Standard JSON Batch
                allResults = await processFileBatch(
                    inputs,
                    options.repair,
                    null, // validateModel - in JS nicht unterstützt
                    options.workers
                );
            }
            
            // Ausgabe
            const successful = allResults.filter(r => r.success).length;
            const failed = allResults.length - successful;
            
            if (options.summary) {
                console.log(`Processed: ${allResults.length}`);
                console.log(`Successful: ${successful}`);
                console.log(`Failed: ${failed}`);
            } else {
                for (const result of allResults) {
                    if (result.success) {
                        console.log(JSON.stringify(result.data));
                    } else {
                        console.error(`ERROR [${result.source}]: ${result.error}`);
                    }
                }
            }
            
            // Optional: JSONL Output
            if (options.output) {
                writeJsonl(allResults, options.output, false);
                console.error(`\nResults written to: ${options.output}`);
            }
            
            // Exit code
            process.exit(failed === 0 ? 0 : 1);
        });
    
    await program.parseAsync();
}

if (require.main === module) {
    main().catch(err => {
        console.error(err);
        process.exit(1);
    });
}

module.exports = {
    BatchResult,
    JSONProcessingError,
    parseJson,
    parseAndValidate,
    processBatch,
    processFileBatch,
    processJsonlFile,
    writeJsonl,
    readJsonl
};

File: perl5/json_batch_processor.pl
#!/usr/bin/perl
# json_batch_processor.py — portiert nach perl5
# Quelle: python, OpenClaw@gateway1:skills/json-utils/scripts/json_batch_processor.py
# auch in: OpenClaw@gateway2:skills/json-utils/scripts/json_batch_processor.py
# Erzeugt: 2026-08-19 durch ABSTRACTIONS_MANAGER.py

use strict;
use warnings;
use utf8;
use JSON;
use File::Basename;
use Getopt::Long;
use Pod::Usage;

# Prüfe ob JSON::XS oder Cpanel::JSON::XS verfügbar ist
my $json_parser;
eval {
    require Cpanel::JSON::XS;
    $json_parser = Cpanel::JSON::XS->new->utf8;
};
if ($@) {
    eval {
        require JSON::XS;
        $json_parser = JSON::XS->new->utf8;
    };
    if ($@) {
        $json_parser = JSON->new->utf8;
    }
}

# Globale Variablen
my $HAS_JSON_PROCESSOR = 0;
my %JSON_PROCESSOR_FUNCS;

# Versuche json_processor.pm zu laden
eval {
    require json_processor;
    json_processor->import();
    $HAS_JSON_PROCESSOR = 1;
    # Dummy-Funktionen falls json_processor nicht geladen werden kann
    *parse_json = sub {
        my ($content, $repair) = @_;
        eval {
            return $json_parser->decode($content);
        } or do {
            die "JSON decode error: $@";
        };
    };
    *parse_and_validate = sub {
        my ($content, $model, $repair) = @_;
        return parse_json($content, $repair);
    };
    *JSONProcessingError = sub {
        my $msg = shift // "";
        return "JSONProcessingError: $msg";
    };
};

# BatchResult Klasse
package BatchResult {
    sub new {
        my ($class, %args) = @_;
        my $self = {
            index => $args{index},
            source => $args{source},
            success => $args{success},
            data => $args{data},
            error => $args{error}
        };
        bless $self, $class;
        return $self;
    }

    sub to_dict {
        my ($self) = @_;
        return {
            index => $self->{index},
            source => $self->{source},
            success => $self->{success},
            data => $self->{data},
            error => $self->{error}
        };
    }
}

# Hauptpaket
package main;

# Liest JSON-Lines (NDJSON) Datei Zeile für Zeile
sub read_jsonl {
    my ($file_path) = @_;
    my @results;
    open my $fh, '<:encoding(UTF-8)', $file_path or die "Cannot open $file_path: $!";
    my $line_num = 0;
    while (my $line = <$fh>) {
        $line_num++;
        chomp $line;
        $line =~ s/^\s+|\s+$//g; # strip
        next if !$line;
        eval {
            my $data = $json_parser->decode($line);
            push @results, { data => $data, line_num => $line_num };
        } or do {
            my $err = $@;
            my $result = BatchResult->new(
                index => $line_num,
                source => "$file_path:$line_num",
                success => 0,
                error => "JSON decode error: $err"
            );
            push @results, $result;
        };
    }
    close $fh;
    return @results;
}

# Verarbeitet eine Liste von Inputs parallel (simuliert durch sequentielle Verarbeitung in Perl)
sub process_batch {
    my ($inputs_ref, $processor_ref, $max_workers) = @_;
    my @results;
    for my $idx (0 .. $#{$inputs_ref}) {
        my $inp = $inputs_ref->[$idx];
        eval {
            my $result = $processor_ref->($inp, $idx);
            push @results, $result;
        } or do {
            my $err = $@;
            my $result = BatchResult->new(
                index => $idx,
                source => "$inp",
                success => 0,
                error => "Unexpected error: $err"
            );
            push @results, $result;
        };
    }
    @results = sort { $a->{index} <=> $b->{index} } @results;
    return @results;
}

# Verarbeitet mehrere JSON-Dateien im Batch
sub process_file_batch {
    my ($file_paths_ref, $repair, $validate_model, $max_workers) = @_;
    
    my $processor = sub {
        my ($path, $idx) = @_;
        eval {
            open my $fh, '<:encoding(UTF-8)', $path or die "Cannot open $path: $!";
            my $content = do { local $/; <$fh> };
            close $fh;
            
            my $data;
            if ($validate_model && $HAS_JSON_PROCESSOR) {
                $data = parse_and_validate($content, $validate_model, $repair);
            } else {
                $data = parse_json($content, $repair);
            }
            
            return BatchResult->new(
                index => $idx,
                source => "$path",
                success => 1,
                data => $data
            );
        } or do {
            my $err = $@;
            return BatchResult->new(
                index => $idx,
                source => "$path",
                success => 0,
                error => "$err"
            );
        };
    };
    
    return process_batch($file_paths_ref, $processor, $max_workers);
}

# Verarbeitet eine JSON-Lines Datei
sub process_jsonl_file {
    my ($file_path, $repair, $validate_model) = @_;
    my @results;
    open my $fh, '<:encoding(UTF-8)', $file_path or die "Cannot open $file_path: $!";
    my $line_num = 0;
    while (my $line = <$fh>) {
        $line_num++;
        chomp $line;
        $line =~ s/^\s+|\s+$//g; # strip
        next if !$line;
        
        eval {
            my $data;
            if ($validate_model && $HAS_JSON_PROCESSOR) {
                $data = parse_and_validate($line, $validate_model, $repair);
            } else {
                $data = parse_json($line, $repair);
            }
            
            push @results, BatchResult->new(
                index => $line_num,
                source => "$file_path:$line_num",
                success => 1,
                data => $data
            );
        } or do {
            my $err = $@;
            push @results, BatchResult->new(
                index => $line_num,
                source => "$file_path:$line_num",
                success => 0,
                error => "$err"
            );
        };
    }
    close $fh;
    return @results;
}

# Schreibt BatchResult-Liste als JSON-Lines
sub write_jsonl {
    my ($results_ref, $output_path, $only_successful) = @_;
    open my $fh, '>:encoding(UTF-8)', $output_path or die "Cannot write to $output_path: $!";
    for my $result (@$results_ref) {
        if ($only_successful && !$result->{success}) {
            next;
        }
        my $dict = $result->to_dict();
        print $fh $json_parser->encode($dict) . "\n";
    }
    close $fh;
}

# Hauptfunktion
sub main {
    my @inputs = @ARGV;
    my $jsonl = 0;
    my $repair = 1;
    my $workers = 4;
    my $output = '';
    my $summary = 0;
    my $help = 0;

    GetOptions(
        'jsonl|l' => \$jsonl,
        'repair|r' => \$repair,
        'workers|w=i' => \$workers,
        'output|o=s' => \$output,
        'summary|s' => \$summary,
        'help|h' => \$help
    ) or pod2usage(2);

    pod2usage(1) if $help;
    pod2usage(2) if !@inputs;

    my @all_results;

    if ($jsonl) {
        # JSON-Lines Modus
        for my $input_path (@inputs) {
            my @results = process_jsonl_file($input_path, $repair);
            push @all_results, @results;
        }
    } else {
        # Standard JSON Batch
        my @file_paths = @inputs;
        @all_results = process_file_batch(\@file_paths, $repair, undef, $workers);
    }

    # Ausgabe
    my $successful = grep { $_->{success} } @all_results;
    my $failed = scalar(@all_results) - $successful;

    if ($summary) {
        print "Processed: " . scalar(@all_results) . "\n";
        print "Successful: $successful\n";
        print "Failed: $failed\n";
    } else {
        for my $result (@all_results) {
            if ($result->{success}) {
                print $json_parser->encode($result->{data}) . "\n";
            } else {
                print STDERR "ERROR [$result->{source}]: $result->{error}\n";
            }
        }
    }

    # Optional: JSONL Output
    if ($output) {
        write_jsonl(\@all_results, $output, 0);
        print STDERR "\nResults written to: $output\n";
    }

    exit($failed == 0 ? 0 : 1);
}

main() unless caller;

__END__

=head1 NAME

json_batch_processor - Batch JSON Processor - Verarbeitet mehrere JSON-Dateien oder JSON-Lines (NDJSON)

=head1 SYNOPSIS

json_batch_processor [options] file1.json [file2.json ...]

 Options:
   --jsonl, -l           Treat inputs as JSON-Lines files
   --repair, -r          Enable JSON repair (default)
   --workers, -w N       Parallel workers (default: 4)
   --output, -o FILE     Output JSON-Lines file
   --summary, -s         Show summary only
   --help, -h            Show this help message

=head1 DESCRIPTION

Dieses Skript verarbeitet mehrere JSON-Dateien oder JSON-Lines-Dateien (NDJSON) und gibt die Ergebnisse aus.

=cut
